src/main/java/com/pinterest/secor/reader/SecorKafkaMessageIterator.java

```diff
@@ -0,0 +1,166 @@
+            public void onPartitionsAssigned(Collection<TopicPartition> collection) {
+                if (skipZookeeperOffsetSeek) {
+                    LOG.debug("offset storage set to kafka. Skipping reading offsets from zookeeper");
+                    return;
+                }
+                Map<TopicPartition, Long> committedOffsets = getCommittedOffsets(collection);
+                committedOffsets.forEach(((topicPartition, offset) -> {
+                    if (offset == -1) {
+                        if (offsetResetConfig.equals("earliest")) {
+                            mKafkaConsumer.seekToBeginning(Collections.singleton(topicPartition));
+                        } else if (offsetResetConfig.equals("latest")) {
+                            mKafkaConsumer.seekToEnd(Collections.singleton(topicPartition));
+                        }
+                    } else {
+                        LOG.debug("Seeking {} to offset {}", topicPartition, Math.max(0, offset));
+                        mKafkaConsumer.seek(topicPartition, Math.max(0, offset));
+                    }
+                }));
+            }
+        };
+
+        String[] subscribeList = config.getKafkaTopicList();
+        if (Strings.isNullOrEmpty(subscribeList[0])) {
+            mKafkaConsumer.subscribe(Pattern.compile(config.getKafkaTopicFilter()), reBalanceListener);
+        } else {
+            mKafkaConsumer.subscribe(Arrays.asList(subscribeList), reBalanceListener);
+        }
+    }
+
+    private Map<TopicPartition, Long> getCommittedOffsets(Collection<TopicPartition> assignment) {
+        Map<TopicPartition, Long> committedOffsets = new HashMap<>();
+
+        for (TopicPartition topicPartition : assignment) {
+            com.pinterest.secor.common.TopicPartition secorTopicPartition =
+                    new com.pinterest.secor.common.TopicPartition(topicPartition.topic(), topicPartition.partition());
+            try {
+                long committedOffset = mZookeeperConnector.getCommittedOffsetCount(secorTopicPartition);
+                committedOffsets.put(topicPartition, committedOffset);
+            } catch (Exception e) {
+                LOG.trace("Unable to fetch committed offsets from zookeeper", e);
+                throw new RuntimeException(e);
+            }
+        }
+
+        return committedOffsets;
+    }
+
+    @Override
+    public void commit(com.pinterest.secor.common.TopicPartition topicPartition, long offset) {
+        TopicPartition kafkaTopicPartition = new TopicPartition(topicPartition.getTopic(), topicPartition.getPartition());
+        OffsetAndMetadata offsetAndMetadata = new OffsetAndMetadata(offset);
+
+        try {
+            LOG.info("committing {} offset {} to kafka", topicPartition, offset);
+            mKafkaConsumer.commitSync(ImmutableMap.of(kafkaTopicPartition, offsetAndMetadata));
+        } catch (CommitFailedException e) {
+            LOG.trace("kafka commit failed due to group re-balance", e);
```

Comment:
Should this exception be rethrown? What if it can't commit the offset?

Reply:
https://github.com/pinterest/secor/blob/master/src/main/java/com/pinterest/secor/reader/LegacyKafkaMessageIterator.java#L77

`LegacyKafkaMessageIterator` - doesn't swallow this exception.

Reply:
I'm pretty sure that we don't need to re-throw it. At this point, the offset was committed to zookeeper and the partition is assigned to another consumer instance. If the consumer is configured to read offsets from kafka, it'll read already uploaded offsets but they'll be skipped since zookeeper commit was successful. If the consumer was configured to read offsets from zookeeper, it will simply continue from there.  

`LegacyKafkaMessageIterator` does not swallow it because it's another lib. I'm not that familiar with the old API, but I believe that it does not have an equivalent exception to `CommitFailedException`. Unfortunately, the exception name can be very misleading since this is not a generic error. It only happens when the commit fails and cannot be retried (group rebalance).